test/testframe.c

```diff
@@ -332,149 +424,181 @@ TestParseCmdLine(int argc, char *argv[])
-                            exit(EXIT_FAILURE);
-                        }
+                if (ret != 0) {
+                    fprintf(stderr, "Error joining thread %d\n", i);
+                    return FAIL;
                 }
+            }
 
-                for (int i = 0; i < max_num_threads; i++) {
-                        ret = pthread_join(threads[i], NULL);
+            free(threads);
+            threads = NULL;
+            free(thread_args);
+            thread_args = NULL;
 
-                        if (ret != 0) {
-                            fprintf(stderr, "Error joining thread %d\n", i);
-                            exit(EXIT_FAILURE);
-                        }
-                    }
-                
+            TestAlarmOff();
 
-                free(threads);
-                free(thread_args);
-                
-                TestAlarmOff();
+            test_num_errs = H5_ATOMIC_LOAD(TestArray[Loop].TestNumErrors);
+            H5_ATOMIC_STORE(TestArray[Loop].TestNumErrors, TestNumErrs_g - test_num_errs);
 
-                test_num_errs = H5_ATOMIC_LOAD(Test[Loop].NumErrors);
-                H5_ATOMIC_STORE(Test[Loop].NumErrors, num_errs_g - test_num_errs);
-                MESSAGE(5, ("===============================================\n"));
-                MESSAGE(5, ("There were %d errors detected.\n\n", (int)H5_ATOMIC_LOAD(Test[Loop].NumErrors)));
+            MESSAGE(5, ("===============================================\n"));
+            MESSAGE(5, ("There were %d errors detected.\n\n", H5_ATOMIC_LOAD(TestArray[Loop].TestNumErrors)));
 #endif /* H5_HAVE_MULTITHREAD */
-            }
         }
     }
 
-    Test_parameters = NULL; /* clear it. */
-
-    if (num_errs_g)
-        print_func("!!! %d Error(s) were detected !!!\n\n", (int)num_errs_g);
+    MESSAGE(2, ("\n\n"));
+    if (TestNumErrs_g)
+        MESSAGE(VERBO_NONE, ("!!! %d Error(s) were detected !!!\n\n", TestNumErrs_g));
     else
         MESSAGE(VERBO_NONE, ("All tests were successful. \n\n"));
+
+    return SUCCEED;
 }
 
 #ifdef H5_HAVE_MULTITHREAD
 /*
  * Set up and execute a test flagged for multi-threaded
  *   execution within a single thread.
  */
-void *ThreadTestWrapper(void *test)
+static void *
+ThreadTestWrapper(void *test)
 {
-    TestCall test_call;
-    int thread_idx;
+    TestStruct *test_struct;
+    int         thread_idx;
 
     assert(test);
 
-    thread_idx = ((TestThreadArgs *)test)->ThreadIndex;
-    test_call = ((TestThreadArgs *)test)->Call;
+    thread_idx  = ((TestThreadArgs *)test)->ThreadIndex;
+    test_struct = ((TestThreadArgs *)test)->Test;
     
-    if (H5_mt_test_thread_setup((int)thread_idx) < 0) {
+    if (H5_mt_test_thread_setup(thread_idx) < 0) {
         fprintf(stderr, "Error setting up thread-local test info");
         return (void*)-1;
     }
 
-    test_call();
+    /* TODO: Test setup and cleanup functions should be
+     * accounted for in the multithread case
+     */
```

Comment:
@mattjala I hadn't yet considered the multi-thread implications of calling the test setup and cleanup callbacks here, so I left this as a TODO. The non-MT case in `PerformTests` above simply looks like

```C
if (TestArray[Loop].TestSetupFunc)
    TestArray[Loop].TestSetupFunc(TestArray[Loop].TestParameters);

TestArray[Loop].TestFunc(TestArray[Loop].TestParameters);

if (TestArray[Loop].TestCleanupFunc)
    TestArray[Loop].TestCleanupFunc(TestArray[Loop].TestParameters);
```

but the multi-thread case may complicate matters.

Reply:
The only tests that delegate threading to testframe are the API tests and the MT VL/testmthdf5 tests. The API tests should have independent resources from threadlocal filenames, and none of the new tests have setup routines. So executing setup and cleanup per-thread should be fine.